readthedocs/doc_builder/environments.py

```diff
@@ -537,28 +548,40 @@ def __enter__(self):
 
     def __exit__(self, exc_type, exc_value, tb):
         """End of environment context"""
-        ret = self.handle_exception(exc_type, exc_value, tb)
+        try:
+            # Update buildenv state given any container error states first
+            self.update_build_from_container_state()
 
-        # Update buildenv state given any container error states first
-        self.update_build_from_container_state()
+            client = self.get_client()
+            try:
+                client.kill(self.container_id)
+            except DockerAPIError:
+                pass
+            try:
+                log.info('Removing container %s', self.container_id)
+                client.remove_container(self.container_id)
+            # Catch direct failures from Docker API, but also requests exceptions
+            # with the HTTP request. These should not
+            except (DockerAPIError, ConnectionError):
+                log.exception(
+                    LOG_TEMPLATE
+                    .format(
+                        project=self.project.slug,
+                        version=self.version.slug,
+                        msg="Couldn't remove container",
+                    ),
+                )
+            self.container = None
+        except BuildEnvironmentError:
```

Comment:
I'm not following this. I think it's not possible to `BuildEnvironmentError` be raised in the `try:` block here. What could be the case?

Reply:
`get_client` raises a `BuildEnvironmentError` on failure here. I'll rethink refactoring this as well though, perhaps a more specific exception, or allowing the docker exception to bubble up makes more sense.